### PR TITLE
Storing Messages, Conversations, Beginning Web Portal

### PR DESCRIPTION
**Pull request by davidgarry1**
The web portal functionality is complete. There are still some issues under the hood to fix, such as SQL injection and storing sensitive info in PHP sessions instead of in javascript. It also needs an aesthetic revamp, which is what the last sprint will be for @david.

@david completed the web portal backend and continues with the frontend
@shuaih worked on the the Java backend
@jonathanloganmoran has been working on the java app GUI

**Bugs**

The toString and fromString methods are broken in the given code for UUIDs. This was a known issue that was supposedly going to be fixed at some point. It caused an issue with adding messages from the database with persistent UUIDs which is a necessary condition.

Since there is a lag between pressing a button and effecting changes to the database (on both the browser and java versions), there will be concurency issues if someone tries to, for instance, add an account with the same name at the same time on the 2 different platforms. This stems from the obvious issue of querying a database to act like a messaging app.

There are some security issues on the website. Besides the aforementioned problems, there is also a problem with scale. If there are 1,000,000 users, for instance, the load time could be as much as 8 minutes, in the way that it is currently implemented. Would be easily fixed but would remove what we consider to be cool features for a small scale messaging app, like seeing all the users and all the conversations.

**Video**

https://www.youtube.com/watch?v=eE7l-dqvv14

**Feature Doc**

https://docs.google.com/document/d/1aqJ1hcsP97f8xbJAh3K6rSVm94bgOvg3J9bJPj9xbXE/edit#

### CODE REVIEW

**Comment by shuai9532** [comment]
New Methods such as AddConversation(),AddMessage(), getMessages() and getConversations() are added. These are the potential methods that I think we might need when communicating with the database. However, we have not yet talked about how we are going to use these methods, meaning what is the return and parameter needed (or I actually need to decide myself before my group mates use it? I am not too sure), and thus I chose the whatever possible return type and parameter type.

**Comment by shuai9532** [comment]
@davidgarry1  my bad, i forgot that the accounts are now connected with conversations and messages. I did not test the drop() while testing conversation and messages, assuming that drop is working. I should have tested drop again.

File: make.sh [hunk 1]
```diff
@@ -15,6 +15,5 @@
 # limitations under the License.
 
 mkdir -p bin
-
-javac -Xlint $(find * | grep "\\.java$") -d ./bin -sourcepath ./src -cp ./third_party/junit4.jar:./bin
-javac -Xlint $(find * | grep "\\.java$") -d ./bin -sourcepath ./test -cp ./third_party/junit4.jar:./bin
+javac -Xlint $(find * | grep "\\.java$") -d ./bin -sourcepath ./src -cp ./third_party/junit4.jar:./third_party/commons-dbcp2-2.1.1.jar:./third_party/mysql-connector-java-5.1.41-bin.jar:./third_party/commons-pool2-2.4.2.jar:./third_party/commons-logging-1.2.jar:./bin
+javac -Xlint $(find * | grep "\\.java$") -d ./bin -sourcepath ./test -cp ./third_party/junit4.jar:./third_party/commons-dbcp2-2.1.1.jar:./third_party/mysql-connector-java-5.1.41-bin.jar:./third_party/commons-pool2-2.4.2.jar:./third_party/commons-logging-1.2.jar:./bin
```
Comment (dentearl) [inline make.sh]: End files with newline character.

File: src/codeu/chat/client/ClientUser.java [hunk 2]
```diff
@@ -150,4 +161,4 @@ public String showUserInfo(String uname) {
   public static void printUser(User user) {
     System.out.println(getUserInfoString(user));
   }
-}
+}
```
Comment (dentearl) [inline src/codeu/chat/client/ClientUser.java]: newline

File: src/codeu/chat/client/simplegui/UserPanel.java [hunk 3]
```diff
@@ -139,33 +158,98 @@ private void initialize() {
...
     userAddButton.addActionListener(new ActionListener() {
       @Override
       public void actionPerformed(ActionEvent e) {
-        final String s = (String) JOptionPane.showInputDialog(
-            UserPanel.this, "Enter user name:", "Add User", JOptionPane.PLAIN_MESSAGE,
-            null, null, "");
-        if (s != null && s.length() > 0) {
-          clientContext.user.addUser(s);
-          UserPanel.this.getAllUsers(listModel);
+        JTextField username = new JTextField();
```
Comment (dentearl) [inline src/codeu/chat/client/simplegui/UserPanel.java]: This block is really similar to the block above, (line 183), is there a way of generalizing the two into a static method and calling that method twice?

File: src/codeu/chat/util/TextValidator.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+  private static final Pattern VALID_USER_NAME_PATTERN = Pattern.compile("[a-zA-Z0-9]+");
+  private static final Pattern VALID_PASSWORD_PATTERN = Pattern.compile("[a-zA-Z0-9]+");
+
+  /**
+   * @param username – the string to be validated
+   * @return – whether or not the string was valid for username type input
+   */
+  public static boolean isValidUserName(String username) {
+    if (username == null) {
```
Comment (dentearl) [inline src/codeu/chat/util/TextValidator.java:19]: This would be a fine place for `com.google.common.base.Strings.isNullOrEmpty()`

Comment (dentearl) [reply src/codeu/chat/util/TextValidator.java:19]: This method and `isValidPassword` are really just special cases of a `private static` method you haven't written yet that takes a `(String textToValidate, Pattern patternToValidate)`

File: src/codeu/scripts/history.sql [hunk 5]
```diff
@@ -0,0 +1,57 @@
+4.14
```
Comment (dentearl) [inline src/codeu/scripts/history.sql:8]: Does this file remain mostly the same? If so, then can you put a comment at the top for readers? If that's not possible / not how it's generated then that's OK. But if it _is_ possible, please do so :)

File: src/database/junitTest/TestConnector.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  @Test
+  public void testGetConversation(){
+    conn.addConversation("323","1238","conversation3");
+    List<String> conversationList = conn.getConversations();
+    for(String str : conversationList){
+      System.out.println(str);
+    }
+    assertFalse(conversationList.isEmpty());
+  }
+}
```
Comment (dentearl) [inline src/database/junitTest/TestConnector.java]: Add a couple round-trip tests. That is, I want to see you have some `private static final UUID = ...` and one for `USER_ID` and `MESSAGE_TITLE` and so on such that you re-use them through all tests. Then, in a few tests at the end, make sure you can `addConversation` and then retrieve that very same conversation, checking for `String.equals()`.

File: test/codeu/chat/TestRunner.java [hunk 7]
```diff
@@ -17,22 +17,89 @@
...
+  private static int FRAME_SPEED = 60;
+
+  /**
+   * updated to a different thread as not to interfere
+   * with the loading message animation
+   *
+   * @param msg: the message to be displayed in the loading
+   */
+  public static synchronized void main(String[] msg) {
+    Thread th = new Thread() {
```
Comment (dentearl) [inline test/codeu/chat/TestRunner.java]: `Thread thread = new Thread() {` please for better readability through this lengy-ish method.

File: test/codeu/chat/TestRunner.java [hunk 8]
```diff
@@ -17,22 +17,89 @@
...
+            numFailures++;
+          }
+          // Print the results of the test.
+          String results = "\rResult:                     ";
+          if(result.wasSuccessful()){
+            results += "\n\tPassed\n";
+          } else {
+            results += "\n\tFailed " + numFailures + " test(s):\n";
+          }
+          System.out.write(results.getBytes());
```
Comment (dentearl) [inline test/codeu/chat/TestRunner.java:65]: `LOG` instead of `System.out.write()`

File: test/codeu/chat/TestRunner.java [hunk 9]
```diff
@@ -17,22 +17,89 @@
...
+   * @param msg: the message to be displayed in the loading
+   */
+  private static synchronized void loading(String msg) {
+    Thread th = new Thread() {
+      @Override
+      public void run() {
+        try {
+          System.out.write("\r|".getBytes());
+          int x =0;
+          while(loading) {
```
Comment (dentearl) [inline test/codeu/chat/TestRunner.java]: Space between `while` and its parenthesis.

File: test/codeu/chat/server/DatabaseTest.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+import database.Connector;
+import java.util.Random;
+import java.util.UUID;
+import org.junit.Test;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+
+
+//Created to test the implementation of the methods in database.Connector.java
```
Comment (dentearl) [inline test/codeu/chat/server/DatabaseTest.java]: Yes! Tests!

(Space after `//`, period after `database.Connector.java`)

File: test/codeu/chat/server/DatabaseTest.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+  private static final database.Connector con = new database.Connector();
+  private static final RandomUuidGenerator uuidGenerator = new RandomUuidGenerator(
+      Uuids.NULL, System.currentTimeMillis());
+
+    @Test
+    public void testAddAndDeleteUser() {
+
+      int iterationsToTest = 6;
+      for(int i=0; i<iterationsToTest; i++) {
```
Comment (dentearl) [inline test/codeu/chat/server/DatabaseTest.java]: Style stuff throughout. Spaces between binary operators.

File: test/codeu/chat/server/DatabaseTest.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        assertTrue(con.addAccount("user"+i, "password", "12343431"+i));
+      }
+
```
Comment (dentearl) [inline test/codeu/chat/server/DatabaseTest.java]: `assertTrue(con.getAllUsers().size() == iterationsToTest);`

File: test/codeu/chat/server/DatabaseTest.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+      assertFalse(con.addAccount(("user"+(iterationsToTest-1)), "password", "12343431"+(iterationsToTest-1)));
```
Comment (dentearl) [inline test/codeu/chat/server/DatabaseTest.java]: Comment here describing what's being tested.

File: test/codeu/chat/server/DatabaseTest.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+      for(int i=0; i<iterationsToTest; i++) {
+        assertTrue(con.deleteAccount("user"+i));
+      }
+
```
Comment (dentearl) [inline test/codeu/chat/server/DatabaseTest.java]: `assertTrue(...` something to the effect that there are no more users.

File: src/codeu/chat/util/TextValidator.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+  public static boolean isValidPassword(String password) {
+    if (password == null) {
+      return false;
+    } else if (password.length() != 0 && VALID_PASSWORD_PATTERN.matcher(password).matches()) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+}
```
Comment (dentearl) [inline src/codeu/chat/util/TextValidator.java]: newline

File: src/codeu/chat/util/store/StoreAccessor.java [hunk 16]
```diff
@@ -1,3 +1,5 @@
+package codeu.chat.util.store;
```
Comment (dentearl) [inline src/codeu/chat/util/store/StoreAccessor.java:1]: Did your intelli-j go nuts on this file or something? ;)

Sorta hilarious how confused github gets by the simple change in indenting, huh?

File: src/database/Connector.java [hunk 17]
```diff
@@ -0,0 +1,477 @@
...
+
+  /* SQL Message Queries */
+  private static final String SQL_INSERT_MESSAGE = String
+      .format("INSERT INTO %s (Uuid, content, id_user, id_conversation) VALUES(?,?,?,?)", MESSAGE);
+  private static final String SQL_SELECT_MESSAGES = String.format(
+      "SELECT * FROM %s WHERE id_conversation = ? ORDER BY creation_time ASC", MESSAGE);
+
+  /* Encryption */
+  private static final Random ram = new SecureRandom();
+  static final private char[] CHARS = "1234567890abcdefghijklmnopqrstuvwxyz-=1,./;'[]".toCharArray();
```
Comment (dentearl) [inline src/database/Connector.java]: keyword ordering should be `private static final`

File: src/database/Connector.java [hunk 18]
```diff
@@ -0,0 +1,477 @@
...
+      "SELECT * FROM %s ORDER BY creation_time ASC", CONVERSATION);
+
+  /* SQL Message Queries */
+  private static final String SQL_INSERT_MESSAGE = String
+      .format("INSERT INTO %s (Uuid, content, id_user, id_conversation) VALUES(?,?,?,?)", MESSAGE);
+  private static final String SQL_SELECT_MESSAGES = String.format(
+      "SELECT * FROM %s WHERE id_conversation = ? ORDER BY creation_time ASC", MESSAGE);
+
+  /* Encryption */
+  private static final Random ram = new SecureRandom();
```
Comment (dentearl) [inline src/database/Connector.java]: I suggest `rand` as a better short name.

File: src/database/Connector.java [hunk 19]
```diff
@@ -0,0 +1,477 @@
...
+  /**
+   * Add conversation into the table in database
+   * @param uuid id of the conversation
+   * @param user_id user who owns the conversation
+   * @param title the name of the conversation, unique
+   * @return true if the conversation is added to the table
+   */
+  public synchronized boolean addConversation(String uuid, String user_id, String title){
+
+    try(Connection conn = ds.getConnection()) {
```
Comment (dentearl) [inline src/database/Connector.java]: Style: space between `try` and first parenthesis, spaces after commas.

File: src/database/Connector.java [hunk 20]
```diff
@@ -0,0 +1,477 @@
...
+      while(result.next()){
+        System.out.println(result.getTimestamp("creation_time").getTime());
+      }
+    } catch (SQLException e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Add conversation into the table in database
```
Comment (dentearl) [inline src/database/Connector.java]: Great use of javadoc.

File: src/database/Connector.java [hunk 21]
```diff
@@ -0,0 +1,477 @@
...
+import org.apache.commons.dbcp2.BasicDataSource;
+
+/**
+ * JDBC connector is needed to connects java program to the database.
+ * Description:
+ * this is the database connector you will need to use to communicate with
+ * the databased hosted on a remote machine.
+ */
+
+public class Connector {
```
Comment (dentearl) [inline src/database/Connector.java]: Overall this class is waaaaay cleaner than the first time I saw it, excellent progress.

File: src/database/Connector.java [hunk 22]
```diff
@@ -0,0 +1,477 @@
...
+        getMessage.setString(1,conversation_id);
+        ResultSet messages = getMessage.executeQuery();
+        while(messages.next()){
+          messageList.add(messages.getString("content"));
+        }
+        return messageList;
+      }
+    }
+    catch (SQLException e) {
+        return messageList;
```
Comment (dentearl) [inline src/database/Connector.java]: indent level

File: src/database/Connector.java [hunk 23]
```diff
@@ -0,0 +1,477 @@
...
+ * JDBC connector is needed to connects java program to the database.
+ * Description:
+ * this is the database connector you will need to use to communicate with
+ * the databased hosted on a remote machine.
+ */
+
+public class Connector {
+
+  /* Database */
+  private BasicDataSource ds;
```
Comment (dentearl) [inline src/database/Connector.java]: `ds` is too short a name for such an important variable that's used over such a broad context (this is Java afterall!). Maybe `dataSource`?

**Review by dentearl — COMMENTED** [review]
Good progress this cycle and I feel like the code quality has improved a good deal too. 

I'm impressed with the demo video. First that you made a demo video, and second with the progress. There are obviously some rough points but it looks slick so far and I'm really excited to see how you start to refine it.

**Pull request closed without merging**